feat: add idv events to api

ilee2u opened the pull request:
- moved what was in signals.py to a handlers.py (which is what their file should have been called)

<!--

Note: Please refer to the Support Development Guidelines on the wiki page to consider backporting to active releases:
https://openedx.atlassian.net/wiki/spaces/COMM/pages/4248436737/Support+Guidelines+for+active+releases

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly readable.
If the linked information must be private (because it contains secrets), clearly label the link as private.

-->

## Description

Describe what this pull request changes, and why. Include implications for people using this change.
Design decisions and their rationales should be documented in the repo (docstring / ADR), per
[OEP-19](https://open-edx-proposals.readthedocs.io/en/latest/oep-0019-bp-developer-documentation.html), and can be
linked here.

Useful information to include:

- Which edX user roles will this change impact? Common user roles are "Learner", "Course Author",
"Developer", and "Operator".
- Include screenshots for changes to the UI (ideally, both "before" and "after" screenshots, if applicable).
- Provide links to the description of corresponding configuration changes. Remember to correctly annotate these
changes.

## Supporting information

Link to other information about the change, such as Jira issues, GitHub issues, or Discourse discussions.
Be sure to check they are publicly readable, or if not, repeat the information here.

## Testing instructions

Please provide detailed step-by-step instructions for testing this change.

## Deadline

"None" if there's no rush, or provide a specific date or event (and reason) if there is one.

## Other information

Include anything else that will help reviewers and consumers understand the change.

- Does this change depend on other changes elsewhere?
- Any special concerns or limitations? For example: deprecations, migrations, security, or accessibility.
- If your [database migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) can't be rolled back easily.

ilee2u committed:
feat: add idv events to api

- moved what was in signals.py to a handlers.py (which is what their file should have been called)

ilee2u commented:
Note To Self: Figure out how to get the latest version of openedx-events installed

ilee2u committed:
chore: quality

ilee2u committed:
fix: rename test file + imports

ilee2u committed:
fix: change handler reverse url in other tests

ilee2u commented on lms/djangoapps/verify_student/tests/test_handlers.py:
```diff
@@ -132,7 +132,7 @@ def setUp(self):
         self.photo_id_key = 'test+key'
 
-    @patch('lms.djangoapps.verify_student.signals.idv_update_signal.send')
+    @patch('lms.djangoapps.verify_student.handlers.idv_update_signal.send')
```
Note that I had to change this reverse URL as otherwise this test failed

alangsto commented on lms/djangoapps/verify_student/handlers.py:
```diff
@@ -0,0 +1,82 @@
+
+# Signal for emitting IDV submission and review updates
+# providing_args = ["attempt_id", "user_id", "status", "full_name", "profile_name"]
+idv_update_signal = Signal()
```
Should this still stay in the signals.py file?

michaelroytman replied:
You defined this signal in `signals.py`, which is what I assume you want to use here, so you should import that signal instead of redefining it here.

alangsto commented on lms/djangoapps/verify_student/handlers.py:
```diff
@@ -0,0 +1,82 @@
+idv_update_signal = Signal()
+
+
+@receiver(SignalHandler.course_published)
```
As is, these handlers won't be connected without importing the file in https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/verify_student/apps.py

michaelroytman commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -70,11 +76,20 @@ def create_verification_attempt(user: User, name: str, status: str, expiration_d
 
 def update_verification_attempt(
     attempt_id: int,
+    user: User,
```
I understand that you need this value to create the event object, but instead of modifying the function signature and requiring a parameter we don't really need, why don't we grab the `User` object off the `VerificationAttempt` instance? Otherwise, this is a breaking change.

michaelroytman commented on lms/djangoapps/verify_student/api.py:
```diff
@@ -120,6 +136,31 @@ def update_verification_attempt(
             )
             raise VerificationAttemptInvalidStatus
 
+        if status == VerificationAttemptStatus.PENDING:
```
I'd recommend emitting the event _after_ the save has completed. If the save fails, we wouldn't want erroneous events to be emitted.

michaelroytman commented on lms/djangoapps/verify_student/handlers.py:
```diff
@@ -0,0 +1,82 @@
+"""
```
If we're going to refactor the signals, I'd recommend following the guidance in [OEP-49: Django App Patterns](https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0049-django-app-patterns.html#signals) for consistency.

ilee2u replied:
Note: We pinned edx-affirmation temporarily in [this PR](https://github.com/openedx/edx-platform/pull/35481) in order to update one of the [reverse urls in edx-name-affirmation](https://github.com/edx/edx-name-affirmation/blob/af8884a3670d5020da4de7b62e4785652bd9a36d/edx_name_affirmation/apps.py#L29) that will now pull from the new `signals` directory. Once that change is pushed, we will unpin `edx-name-affirmation` in this PR.

ilee2u replied:
> `edx-name-affirmation` in this PR.

edx-name-affirmation has been unpinned.

michaelroytman commented on lms/djangoapps/verify_student/signals.py:
```diff
@@ -1,83 +1,112 @@
-        except ObjectDoesNotExist:
-            VerificationDeadline.set_deadline(course_key, course.end)
+    user_data = UserData(
+        # NOTE to self: the id field was previously = lms_user_id in edx-exams, which I copied this from
```
`User.id` is fine. `lms_user_id` was specific to the `edx-exams` IDA because `User.id` in that application referred to the `id` of the `User` in the `edx-exams` database and not the `id` of the `User` in the `edx-platform` database (which was the purpose of `lms_user_id`).

michaelroytman reviewed (changes requested):
I think we should add tests to verify that the signals are properly emitted. Could you add those, please?

ilee2u committed:
fix: refactor signals and handlers pattern

- following OEP-49 pattern for signals directory
- user removed as param for update function
- event now emitted after save

ilee2u committed:
fix: unpin edx-name-affirmation

ilee2u committed:
chore: add init to signals dir

ilee2u committed:
fix: compile requirements

ilee2u committed:
chore: quality

ilee2u committed:
chore: fix some imports

ilee2u committed:
chore: quality

ilee2u committed:
test: added signal emissions to test_api

ilee2u committed:
chore: lint

schenedx reviewed (approved):
👍

michaelroytman reviewed (approved):
Looks great!
